docs/proposal-007-apply.md

```diff
@@ -0,0 +1,200 @@
+* [Open questions](#open-questions)
+
+## Goals
+
+The purpose of `apply` is to subsume the various imperative commands
+`eksctl` provides now into one `apply` command that reconciles the actual state
+of an EKS cluster with the intended state as specified in the `eksctl` config file.
+
+## Non-goals
+
+This proposal is only concerned with the `apply` subcommand. Existing behavior
+and commands are unaffected. Continuing to support them shouldn't complicate
+the implementation too much as they are more or less subsets of
+operations performed by `apply`.
+
+## Motivation
+
+`eksctl` supplies users with imperative commands for clusters and cluster
+resources that allow them to take explicit actions, like `create` or `delete`.
+We also encourage the use of a config file in which the desired state of the
+cluster is described. Still, users are required to figure out which imperative
+steps to take to reconcile the cluster with the desired state.
+The missing capability is to give `eksctl` the ability to do this.
+Conceptually, `eksctl` would gather all of the information it can about the
+current state of the cluster in order to build a `ClusterConfig` describing
+the real-world cluster and then `diff` this config with the user-provided
+`ClusterConfig`. The `diff` is used to setup a plan of modification to the
+cluster and resources.
+
+### Questions for discussion
+
+This proposal also opens the floor for discussion around non-straightforward
+questions around reconciliation.
+
+#### Deletion of resources
+
+When looking at a list of nodegroups in a cluster, it's not clear how to handle
+deletion.
+
+-   Is the given list intended to be a complete list of nodegroups and
+    no others should exist? Are they automatically deleted?
+-   How do we handle non-eksctl managed resources?
+
+#### Immutability
+
+In order to change some properties, entire resource will have to be recreated.
+This should be explicit to the user.
+
+#### Gradual support and config
+
+Because it's impractical to introduce complete support for reconciliation at
+once, we need to instead gradually support properties and resources of a
+cluster.
+
+This leads to the question of whether the current `ClusterConfig` is a good
+structure for a config meant for reconciliation.
+
+## Proposal
+
+Introduce a subcommand `apply` which has only a config file argument.
+`apply` would reconcile desired with real-world state
+[as described in the motivation](#motivation).
+The exact details per resource will potentially differ depending on
+the resources and won't be enumerated here.
+
+1. `apply` is put behind the explicit "experimental" flag (cf `enable profile/repo`)
+2. Take full ownership of `eksctl`-managed resources (i.e. delete missing resources by default)
+3. Options for non-`eksctl`-managed resources as well as resources where
+   ownership is unknowable are covered below
+4. Changing immutable fields through recreation is the default
```

Comment:
This feels like something that would be quite big piece of work. An example would be disabling OIDC which isn't currently possible, do we also delete all of the IAMServiceAccounts as they wouldn't have been created in the first place if OIDC was false. 

A more explosive example would be if I specified a lower version of Kubernetes in my config would we delete the existing cluster and create a new cluster at the lower version, same for nodegroups etc

Reply:
you can disable OIDC (not through eksctl), but yes all dependent resources need to be altered/recreated. what else can we do? the user would have to fall back to imperative commands otherwise

Reply:
We could have some immutable fields that we don't support changing through reaction. I don't think we should do it lightly, but some fields seem impractical to support.

Reply:
I think in that case the user will then have to just delete and recreate the resources themselves, I think the power of `apply` is that `eksctl` recognizes and handles that case. See [the example of nodegroups, a big motivator for `apply`](https://github.com/weaveworks/eksctl/issues/462) and [a duplicate issue](https://github.com/weaveworks/eksctl/issues/443). I don't think the user should "ever" have to fallback to imperative commands and I don't see any other recourse if they can't replace it. What should the user do if they can't recreate a resource because of a field.

Reply:
I think the `nodegroups` example makes complete sense to support, I'm not disagreeing, very few fields are worth restricting.

To be more specific about whats worth restricting I think config field changes that would invoke the deletion and recreation of the cluster itself are good examples. They feel very explosive and might not be what the user expects. They might not know that downgrading kubernetes version isn't possible on an existing cluster for example, or other fields like modifying the subnets a cluster uses. We do have the `apply --plan` to display that to them before the destruction occurs, is implementing more complex codepaths like this worth? I might be wrong, but it feels like a discussion worth having

Reply:
Yeah I was definitely assuming we have `apply --plan`, for sure. I don't think the codepath needs to be complex, I think we need only know, for each resource, on which other resources it depends.
With CAPA you can't make such changes, but it's running as a controller and there's no great way to handle the "approve" flow with custom resources, at the moment.

Reply:
How much we should consider the case of "`eksctl` as a controller" here I'm definitely not sure about and should be open to discussion IMO!